### PR TITLE
Added Strassen's Algorithm in python

### LINKED ISSUE #695: Adding Strassen's matrix multiplication Algorithm in c,python and java
Strassen's algorithm is used to find product of two matrix with more efficiency than the conventional one.

### PR DESCRIPTION
**Pull request by Nami2012**
Fixes #695 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:

- Implemented Strassen's Algorithm in Python for 2x2 matrix and NxN matrix

#### Languages Used:
- Python

#### Files Added:
- Strassens_Algorithm.py	
- Strassens_Algorithm_Input_Size.py

### CODE REVIEW

**Comment by MastersAbh** [comment]
Hey, you can remove the first file. Code specifically for 2X2 matrix is not required.

**Comment by Nami2012** [comment]
Removed code for 2x2 matrix

File: Strassens_Algorithm/Strassens_Algorithm.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    mid = matrix_length // 2
+    A00 = [[a[i][j] for j in range(mid)] for i in range(mid)]
+    A10 = [[a[i][j] for j in range(mid)] for i in range(mid, matrix_length)]
+    A01 = [[a[i][j] for j in range(mid, matrix_length)] for i in range(mid)]
+    A11 = [[a[i][j] for j in range(mid, matrix_length)] for i in range(mid, matrix_length)]
+    return A00 , A01,A10,A11
+
+#Add submatrices
+def add_matrix(a, b):
+
```
Comment (MastersAbh) [inline Strassens_Algorithm/Strassens_Algorithm.py]: remove extra space

File: Strassens_Algorithm/Strassens_Algorithm.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#Add submatrices
+def add_matrix(a, b):
+        
+    return [[a[i][j] + b[i][j]
+             for j in range(len(a))] for i in range(len(a))]
+
+#Subtract submatrices
+def sub_matrix(a, b):
+
```
Comment (MastersAbh) [inline Strassens_Algorithm/Strassens_Algorithm.py]: extra space

File: Strassens_Algorithm/Strassens_Algorithm.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#Subtract submatrices
+def sub_matrix(a, b):
+        
+    return [[a[i][j] - b[i][j]
+             for j in range(len(a))] for i in range(len(a))]
+
+#Function that implements Strassen's algorithm 
+def Strassen(a,b):
+    c = []
+
```
Comment (MastersAbh) [inline Strassens_Algorithm/Strassens_Algorithm.py]: extra space

File: Strassens_Algorithm/Strassens_Algorithm.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+             for j in range(len(a))] for i in range(len(a))]
+
+#Function that implements Strassen's algorithm 
+def Strassen(a,b):
+    c = []
+    
+    if len(a) == 1 :
+        c.append([])
+        c[0].append(a[0][0] * b[0][0])
+
```
Comment (MastersAbh) [inline Strassens_Algorithm/Strassens_Algorithm.py]: extra space

File: Strassens_Algorithm/Strassens_Algorithm.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    c = []
+    
+    if len(a) == 1 :
+        c.append([])
+        c[0].append(a[0][0] * b[0][0])
+    
+    else :
+        A00,A01,A10,A11 = split_matrix(a)
+        
+
```
Comment (MastersAbh) [inline Strassens_Algorithm/Strassens_Algorithm.py]: extra spaces

File: Strassens_Algorithm/Strassens_Algorithm.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    for j in range(s):
+        n =int(input())
+        a[i].append(n) 
+print(a)
+
+#Read elements of second matrix
+b = [] 
+
+for i in range(s): 
+
```
Comment (MastersAbh) [inline Strassens_Algorithm/Strassens_Algorithm.py]: extra space

File: Strassens_Algorithm/Strassens_Algorithm.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        a[i].append(n) 
+print(a)
+
+#Read elements of second matrix
+b = [] 
+
+for i in range(s): 
+    
+    b.append([]) 
+
```
Comment (MastersAbh) [inline Strassens_Algorithm/Strassens_Algorithm.py]: extra space

File: Strassens_Algorithm/Strassens_Algorithm.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+#Function to form sub matrices
+def split_matrix(a):
+    matrix_length = len(a)
+    mid = matrix_length // 2
+    A00 = [[a[i][j] for j in range(mid)] for i in range(mid)]
+    A10 = [[a[i][j] for j in range(mid)] for i in range(mid, matrix_length)]
+    A01 = [[a[i][j] for j in range(mid, matrix_length)] for i in range(mid)]
+    A11 = [[a[i][j] for j in range(mid, matrix_length)] for i in range(mid, matrix_length)]
+    return A00 , A01,A10,A11
```
Comment (MastersAbh) [inline Strassens_Algorithm/Strassens_Algorithm.py]: please remove spaces around ,

Comment (MastersAbh) [reply Strassens_Algorithm/Strassens_Algorithm.py]: Please resolve this also.

Comment (Nami2012) [reply Strassens_Algorithm/Strassens_Algorithm.py]: can you please specify ?

Comment (MastersAbh) [reply Strassens_Algorithm/Strassens_Algorithm.py]: A00 , A01,A10,A11 should be A00,A01,A10,A11

File: Strassens_Algorithm/Strassens_Algorithm.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
+#Strassen's matrix multiplication Algorithm for NxN matrix
+
+#Function to form sub matrices
+def split_matrix(a):
+    matrix_length = len(a)
+    mid = matrix_length / 2
```
Comment (MastersAbh) [inline Strassens_Algorithm/Strassens_Algorithm.py]: Do not change // to / . mid should be integer to be used as an argument to range()

Comment (Nami2012) [reply Strassens_Algorithm/Strassens_Algorithm.py]: ohh yeah ,I'll correct it. i mistook that for the resolving issue

File: Strassens_Algorithm/Strassens_Algorithm.py (hunk 8 above)
Comment (MastersAbh) [reply Strassens_Algorithm/Strassens_Algorithm.py]: You have to remove spaces around comma here

File: Strassens_Algorithm/Strassens_Algorithm.py (hunk 9 above)
Comment (MastersAbh) [reply Strassens_Algorithm/Strassens_Algorithm.py]: Sure

**Comment by MastersAbh** [comment]
Good work. Last thing. Add introduction about strassen's formula at the top of code and a sample input output at the end of file as comments.
It will be all.

**Comment by Nami2012** [comment]
@MastersAbh @KanchanThareja 
Thanks.Will add the introduction and output.
One last thing , how do I squash commits?Triying it for the first time ,getting too much errors!!

**Comment by MastersAbh** [comment]
Sure, Please connect on slack channel. Will be happy to help. :)

**Comment by Nami2012** [comment]
@MastersAbh @KanchanThareja  Squashed the commits, anymore changes?

**Comment by MastersAbh** [comment]
Good work @Nami2012

**Comment by MastersAbh** [comment]
@jainaman224  It is approved. You can have a final look.

**Pull request merged**